[CALCITE-5644] Add CONTAINS_SUBSTR function (enabled in BigQuery library)

tanclary opened the pull request:
BigQuery's `CONTAIN_SUBSTR` determines whether a string appears in an expression. This expression can be of many types, as seen in the docs [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/string_functions#contains_substr).

Also removes some random extra whitespace. Let me know if this should be reverted.

Any comments, questions, or suggestions in the meantime are appreciated as always.

julianhyde commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -562,6 +564,91 @@ public static List<ByteString> split(ByteString s, ByteString delimiter) {
     }
   }
 
+  public static @Nullable Boolean containsSubstr(@Nullable Object @Nullable [] s0, String s1) {
```
Each of these methods needs a simple javadoc comment. Without such comments, it's quite difficult to map from java types to the SQL types.

julianhyde commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -562,6 +564,91 @@ public static List<ByteString> split(ByteString s, ByteString delimiter) {
+  }
+
+  /** SQL {@code CONTAINS_SUBSTR(string, string)} function. */
+  public static Boolean containsSubstr(@Nullable Object s0, String s1) {
```
Do all these methods need to return `Boolean` rather than `boolean`?

Remember that Calcite's code generator will short-circuit the call based on the function's strong-ness. For normal functions, if any arguments are null, it will not call the function, and the result is SQL NULL (or UNKNOWN). In such cases, you should not handle nulls. `Integer` arguments can be `int`, `Boolean` return can be `boolean`, etc.

tanclary replied:
I left some as `Boolean` because there are cases where the method should return `NULL` even if none of the parameters are `NULL`. For example if the argument is ROW(1, null, 3) and the substr is '2', it should return `NULL`.

julianhyde commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -562,6 +564,91 @@ public static List<ByteString> split(ByteString s, ByteString delimiter) {
+    }
+  }
+
+  /** SQL {@code CONTAINS_SUBSTR(string, string)} function. */
```
Use more descriptive names for arguments. Make the arguments in the javadoc comment (`string`) and Java parameters (`s0`, `s1`) consistent.

tanclary commented on core/src/main/java/org/apache/calcite/sql/fun/SqlContainsSubstrFunction.java:
```diff
@@ -0,0 +1,47 @@
+
+/** The "CONTAINS_SUBSTR(expression, search_value_literal[, json_scope => json_scope_value])"
+ * function; performs a search to see if a value exists as a substring in an expression. */
+public class SqlContainsSubstrFunction extends SqlFunction {
```
Right now if you use `SqlBasicFunction.create`, an `UnsupportedOperationException` gets thrown from`validateColumnListParams` [here](https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L6363). The `deriveType` override avoids this. What do you think?

julianhyde replied:
`validateColumnListParams` seems to always throw. It also seems vestigial. If you remove the one call to it, do any tests break? I suspect not.

So perhaps the method and calls can be removed. If you do that, can you switch to using `SqlBasicFunction.create`?

tanclary commented:
@julianhyde I added a commit that I think addresses most of your comments. Thanks for the thorough review, i think the PR is in much better shape now. Of course let me know if you have more suggestions.

julianhyde commented:
Is the new function class `SqlContainsSubstrFunction` necessary?

Thanks for adding javadoc in SqlFunctions. For future reference, `{@code ... }` is slightly preferred over `<code> ... </code>` because it handles some HTML escaping. (Of course you can't use it if you actually want HTML markup e.g. `<b>`and `<br>`.)

I'd rename `enum JsonScopeValue` to `enum JsonScope`. Every class holds values.

Make `normalize` private.

tanclary commented:
> Is the new function class `SqlContainsSubstrFunction` necessary?
> 
> Thanks for adding javadoc in SqlFunctions. For future reference, `{@code ... }` is slightly preferred over `<code> ... </code>` because it handles some HTML escaping. (Of course you can't use it if you actually want HTML markup e.g. `<b>`and `<br>`.)
> 
> I'd rename `enum JsonScopeValue` to `enum JsonScope`. Every class holds values.
> 
> Make `normalize` private.
I think right now the class is not fully necessary, but I think there are other `CONTAINS` functions (such as `REGEXP_CONTAINS`) that could make having its own dedicated class useful in the future. Additionally, there is currently an exception that gets thrown from `validateColumnListParams` if we just use `SqlBasicFunction.create` instead. The `#deriveType` override avoids this. Let me know what you think.

Good to know about the comments.

I amended to address your 3rd and 4th points.

tanclary commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -562,6 +564,91 @@ public static List<ByteString> split(ByteString s, ByteString delimiter) {
+    LinkedHashMap<String, String> map = (LinkedHashMap) JsonFunctions.dejsonize(s0);
+    Set<String> keys = map.keySet();
+    Collection<String> values = map.values();
+    if (s2.equals("JSON_KEYS")) {
```
Done

tanclary commented on core/src/main/java/org/apache/calcite/sql/SqlKind.java:
```diff
@@ -839,6 +839,8 @@ public enum SqlKind {
    * TABLE(udx(CURSOR(SELECT ...), x, y, z))</code>. */
   CURSOR,
 
+  CONTAINS_SUBSTR,
```
Done

tanclary commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -562,6 +564,91 @@ public static List<ByteString> split(ByteString s, ByteString delimiter) {
+  }
+
+  public static @Nullable Boolean containsSubstr(String s0, String s1, String s2) {
+    // The third argument specifies the json_scope, either keys, values, or both
```
Done

tanclary commented:
@julianhyde I removed the class I added in favor of `SqlBasicFunction.create()` as you suggested, I also removed the method and its lone usage. Let me know what you think.
